spec/rubocop/cop/sorbet/obsolete_strict_memoization_spec.rb

```diff
@@ -0,0 +1,285 @@
+          RUBY
+        end
+      end
+    end
+
+    context "when its not the first line in a method" do
+      it "registers an offence and autocorrects" do
+        expect_offense(<<~RUBY)
+          def foo
+            some
+            other
+            code
+            @foo = T.let(@foo, T.nilable(Foo))
+            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ This two-stage workaround for memoization in `#typed: strict` files is no longer necessary. See https://sorbet.org/docs/type-assertions#put-type-assertions-behind-memoization.
+            @foo ||= Foo.new
+          end
+        RUBY
+
+        expect_correction(<<~RUBY)
+          def foo
+            some
+            other
+            code
+            @foo ||= T.let(Foo.new, T.nilable(Foo))
+          end
+        RUBY
+      end
+    end
+
+    context "using an old version of Sorbet" do
+      # For old versions, the obsolete memoization pattern isn't actually obsolete.
+
+      describe "the obsolete memoization pattern" do
+        it "does not register an offence" do
+          allow(Bundler).to(receive(:locked_gems)).and_return(
+            Struct.new(:specs).new([
+              *specs,
+              Gem::Specification.new("sorbet-static", "0.5.10209"),
+            ]),
+          )
+
+          expect_no_offenses(<<~RUBY)
+            sig { returns(Foo) }
+            def foo
+              @foo = T.let(@foo, T.nilable(Foo))
+              @foo ||= Foo.new
+            end
+          RUBY
+        end
+      end
+    end
+
+    context "not using Sorbet" do
+      # If the project is not using Sorbet, the obsolete memoization pattern might be intentional.
+
+      describe "the obsolete memoization pattern" do
+        it "does not register an offence" do
+          allow(Bundler).to(receive(:locked_gems)).and_return(
+            Struct.new(:specs).new([
+              *specs,
+            ]),
+          )
+
+          expect_no_offenses(<<~RUBY)
+            sig { returns(Foo) }
+            def foo
+              @foo = T.let(@foo, T.nilable(Foo))
+              @foo ||= Foo.new
+            end
+          RUBY
+        end
+      end
+    end
+  end
+
+  describe "a mistaken variant of the obsolete memoization pattern" do
+    it "registers an offence and autocorrects" do
+      # This variant would have been a mistake, which would have caused the memoized value to be discarded
+      # and recomputed on every call. We can fix it up into the working version.
+
+      expect_offense(<<~RUBY)
+        def foo
+          @foo = T.let(nil, T.nilable(Foo))
+          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ This two-stage workaround for memoization in `#typed: strict` files is no longer necessary. See https://sorbet.org/docs/type-assertions#put-type-assertions-behind-memoization.
+          @foo ||= Foo.new
+        end
+      RUBY
+
+      expect_correction(<<~RUBY)
+        def foo
+          @foo ||= T.let(Foo.new, T.nilable(Foo))
+        end
+      RUBY
+    end
+  end
```

Comment:
This correction is unsafe, so I'm extracting it to a new `BuggyObsoleteMemoizationCopy` that's correctly marked `SafeAutoCorrect: false`:

https://github.com/Shopify/rubocop-sorbet/pull/175